### PR TITLE
feat: [AXIMST-344] Copy/paste functionality base

### PR DESCRIPTION
**Pull request by monteri**
https://youtrack.raccoongang.com/agiles/104-1224/105-3378?issue=AXIMST-344

### CODE REVIEW

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import MockAdapter from 'axios-mock-adapter';
+import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+import userEvent from '@testing-library/user-event';
+
+import initializeStore from '../../../../store';
+import ActionButtons from './ActionButtons';
+import { getClipboardUrl, getCourseUnitApiUrl } from '../../../data/api';
+import { clipboardUnit, courseUnitIndexMock } from '../../../__mocks__';
+import { executeThunk } from '../../../../utils';
+import { copyToClipboard, fetchCourseUnitQuery } from '../../../data/thunk';
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx:14]:
```suggestion
import initializeStore from '../../../../store';
import { executeThunk } from '../../../../utils';
import { copyToClipboard, fetchCourseUnitQuery } from '../../../data/thunk';
import { getClipboardUrl, getCourseUnitApiUrl } from '../../../data/api';
import { clipboardUnit, courseUnitIndexMock } from '../../../__mocks__';
import ActionButtons from './ActionButtons';
```

File: src/i18n/messages/fa_IR.json [hunk 2]
```diff
@@ -66,7 +66,7 @@
   "course-authoring.course-unit.xblock.button.edit.alt": "Edit Item",
   "course-authoring.course-unit.xblock.button.copy.label": "Copy",
   "course-authoring.course-unit.xblock.button.duplicate.label": "Duplicate",
-  "course-authoring.course-unit.xblock.button.move.label": "Move",
+  "course-authoring.course-unit.xblock.button.copyToClipboard.label": "Copy to clipboard",
```
Comment (PKulkoRaccoonGang) [inline src/i18n/messages/fa_IR.json:69]: [important]: Please return the translation you deleted

> "course-authoring.course-unit.xblock.button.move.label": "Move",

File: src/course-unit/course-xblock/CourseXBlock.jsx [hunk 3]
```diff
@@ -41,6 +45,11 @@ const CourseXBlock = ({ id, title }) => {
                 <Dropdown.Item>
                   {intl.formatMessage(messages.blockLabelButtonMove)}
                 </Dropdown.Item>
+                {enableCopyPasteUnits && (
+                  <Dropdown.Item onClick={() => dispatch(copyToClipboard(id))}>
```
[the remark this answers is not in the document]
Comment (PKulkoRaccoonGang) [reply src/course-unit/course-xblock/CourseXBlock.jsx:49]: [clarify]: Also now we have two buttons for copying, should it be like this?

**Implementation:**
Screenshot 2024-01-28 at 14 12 54

**Design:**
Screenshot 2024-01-28 at 14 15 28

File: src/course-unit/data/thunk.js [hunk 4]
```diff
@@ -239,3 +246,37 @@ export function fetchCourseVerticalChildrenData(itemId) {
     }
   };
 }
+
+export function copyToClipboard(usageKey) {
+  const POLL_INTERVAL_MS = 1000; // Timeout duration for polling in milliseconds
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/data/thunk.js:251]: [question]: If we plan to use notification delay for smoothness and improve the user experience, should we put this constant in the appropriate constants file?

File: src/course-unit/data/thunk.js [hunk 5]
```diff
@@ -70,6 +76,7 @@ export function fetchCourseSectionVerticalData(courseId, sequenceId) {
         modelType: 'units',
         models: courseSectionVerticalData.units,
       }));
+      dispatch(updateClipboardData(courseSectionVerticalData.userClipboard));
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/data/thunk.js:79]: 👍

File: src/course-unit/data/thunk.js (hunk 4 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/data/thunk.js:251]: [clarify]: After you added a mandatory delay for notifications to show, should I still be able to see the notifications now?

https://github.com/raccoongang/frontend-app-course-authoring/assets/93188219/3ef3a6db-37aa-479b-b781-b331f4e76eeb

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    axiosMock
+      .onGet(getClipboardUrl())
+      .reply(200, clipboardUnit);
+    await executeThunk(fetchCourseUnitQuery(courseId), store.dispatch);
+  });
+
+  it('render ActionButtons component with Copy to clipboard', () => {
+    const { getByTestId } = renderComponent();
+
+    expect(getByTestId('copy-xblock-to-clipboard')).toBeInTheDocument();
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx:61]: [important]: Now you find the button using date-test-id, let's find it using `getByRole` and text content

For example: 
```suggestion
    const copyXBlockBtn = getByRole('button', { name: messages.actionButtonCopyUnitTitle.defaultMessage });
    expect(copyXBlockBtn).toBeInTheDocument();
```

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  });
+  it('click on the Copy to clipboard button updates clipboardData', async () => {
+    const { getByTestId } = renderComponent();
+
+    const copyToClipboardButton = getByTestId('copy-xblock-to-clipboard');
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx:66]: [fiy]: Do not search by ID for elements that can be found by role :)

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    await act(async () => {
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.test.jsx:67]: [important]: Please remove the unnecessary act(...) wrapper, `userEvent` is already wrapped in it

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.jsx [hunk 9]
```diff
@@ -35,9 +38,11 @@ const ActionButtons = () => {
       )}
       {enableCopyPasteUnits && (
         <Button
+          onClick={() => dispatch(copyToClipboard(id))}
           className="mt-2"
           variant="outline-primary"
           size="sm"
+          data-testid="copy-xblock-to-clipboard"
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.jsx:45]: [important]: don't forget to remove `data-testid` after refactoring the tests

File: src/course-unit/sidebar/components/sidebar-footer/ActionButtons.jsx [hunk 10]
```diff
@@ -1,13 +1,16 @@
-import { useSelector } from 'react-redux';
+import { useDispatch, useSelector } from 'react-redux';
 import { Button } from '@edx/paragon';
 import { useIntl } from '@edx/frontend-platform/i18n';
 
 import { getCourseUnitData } from '../../../data/selectors';
 import messages from '../../messages';
+import { copyToClipboard } from '../../../data/thunk';
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/sidebar/components/sidebar-footer/ActionButtons.jsx:7]:
```suggestion
import { copyToClipboard } from '../../../data/thunk';
import { getCourseUnitData } from '../../../data/selectors';
import messages from '../../messages';
```

File: src/course-unit/data/api.js [hunk 11]
```diff
@@ -130,6 +131,27 @@ export async function createCourseXblock({
   return data;
 }
 
+/**
+ * Retrieves user's clipboard.
+ * @returns {Promise<Object>} - A Promise that resolves clipboard data.
+ */
+export async function getClipboard() {
+  const { data } = await getAuthenticatedHttpClient().get(getClipboardUrl());
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/data/api.js:139]: [optional]:
```suggestion
  const { data } = await getAuthenticatedHttpClient()
     .get(getClipboardUrl());
```

File: src/course-unit/data/api.js [hunk 12]
```diff
@@ -130,6 +131,27 @@ export async function createCourseXblock({
...
+  return camelCaseObject(data);
+}
+
+/**
+ * Updates user's clipboard.
+ * @param {string} usageKey - The ID of the block.
+ * @returns {Promise<Object>} - A Promise that resolves clipboard data.
+ */
+export async function updateClipboard(usageKey) {
+  const { data } = await getAuthenticatedHttpClient().post(getClipboardUrl(), { usage_key: usageKey });
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/data/api.js:150]: [optional]:
```suggestion
  const { data } = await getAuthenticatedHttpClient()
    .post(getClipboardUrl(), { usage_key: usageKey });
```

File: src/course-unit/data/thunk.js [hunk 13]
```diff
@@ -239,3 +246,37 @@ export function fetchCourseVerticalChildrenData(itemId) {
...
 }
+
+export function copyToClipboard(usageKey) {
+  const POLL_INTERVAL_MS = 1000; // Timeout duration for polling in milliseconds
+
+  return async (dispatch) => {
+    dispatch(updateClipboardData(null));
+    dispatch(showProcessingNotification(NOTIFICATION_MESSAGES.copying));
+    dispatch(updateSavingStatus({ status: RequestStatus.IN_PROGRESS }));
+    dispatch(updateQueryPendingStatus(true));
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/data/thunk.js:257]: [important]: Currently, the notification about Internet problems does not work as expected (an error message and a message about the lack of Internet connection appear). Note how this works for the functionality of adding a new unit.

https://github.com/raccoongang/frontend-app-course-authoring/assets/93188219/18a94a5e-1ee2-4372-a479-135dc0378dc7

**Review by PKulkoRaccoonGang — CHANGES_REQUESTED** [review]
@monteri The copy functionality looks great! 
Left comments for improvements and bug fixes. 
Please make sure that the waffle flag `contentstore.enable_copy_paste_units` is present in the documentation for testing

File: src/course-unit/clipboard/useClipboard.jsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
+import { useEffect, useState } from 'react';
+import { useSelector } from 'react-redux';
+import { getClipboardData } from '../data/selectors';
+import { NOT_XBLOCK_TYPES, STUDIO_CLIPBOARD_CHANNEL } from '../constants';
```
[the remark this answers is not in the document]
Comment (monteri) [reply src/course-unit/clipboard/useClipboard.jsx:4]: 👍

**Pull request merged**